vac/raw/mix.md

```diff
@@ -0,0 +1,658 @@
+    - TCP/UDP port number (2 bytes)
+    - QUIC/TLS protocol identifier flag (1 byte)
+    - Peer ID (32 bytes for Ed25519 or Secp256k1).
+
+The entire Sphinx packet header ($α$, $β$, and $γ$) can fit within a fixed size
+of $32 + (2r+t)\kappa + 16 = 256$ bytes, leaving ample room for a large $δ$ of
+up to $2413 - 256 = 2157$ bytes.
+
+#### 4.3 Message Format
+
+The Mix protocol uses the Sphinx packet format to encapsulate messages and
+routing information.
+
+```proto
+message SphinxPacket {
+  bytes alpha = 1; // 32 bytes
+  bytes beta = 2; // 192 - 208 bytes
+  bytes gamma = 3; // 16 bytes
+  bytes delta = 4; // variable size, max 2157 bytes
+}
+```
+
+### 5. Handler Function
+
+The handler function is responsible for processing connections and messages for
+the Mix protocol. It operates according to the mix node roles (_i.e.,_ sender,
+intermediary mix node, or exit node) defined in
+[Section 2.1](#21-mix-nodes-roles). This function is crucial for implementing
+the core functionality of the mixnet protocol within the libp2p framework.
+
+When a node receives a new stream for the `"/mix/1.0.0"` protocol, the handler
+function is invoked. It performs different operations based on the node's role
+in the current message path:
+
+- **Role Determination**
+
+  The handler first determines the node's role for the incoming message. This
+  is typically done by examining the packet structure and the node's position
+  in the network.
+
+- **Packet Processing**
+
+  Depending on the role, the handler processes the Sphinx packet differently:
+
+  - For senders, it creates and sends new Sphinx packets.
+  - For intermediary nodes, it processes and forwards existing packets.
+  - For exit nodes, it decrypts the final layer and disseminates the original message.
+
+- **Error Handling**
+
+  It manages any errors that occur during packet processing, such as invalid
+  MACs or decryption failures.
+
+- **Logging and Metrics**
+
+  The handler is also be responsible for logging important events and
+  collecting metrics for network analysis and debugging.
+
+The specific implementation of the handler function for each role (_i.e.,_
+sender, intermediary, and exit node) is detailed in the following subsections.
+
+#### 5.1 Sender
+
+1. **Convert the libp2p Message to Bytes**
+
+   Serialize the libp2p message to bytes and store the result in
+   `libp2p_message`. This can be done using Protocol Buffers or another
+   serialization method.
+
+2. **Calculate Proof of Work (PoW)**
+
+   Perform the following steps to compute the PoW challenge and response:
+
+   i. **Create Challenge**
+
+   Retrieve the current Unix timestamp, `timestamp`, in seconds (4 bytes).
+
+   ii. **Find A Valid Nonce**
+   - Initialize the `nonce` to a 4-byte value (initially zero).
+   - Increment the `nonce` until the SHA-256 hash of
+    `<libp2p_message_bytes | timestamp | nonce>` has at least
+    18 leading zeros.
+   - The final value of the `nonce` is considered valid.
```

Comment:
since most of the projects within the IFT are targeting resource constrained devices, perhaps we should do an analysis on how much time it takes for cheaper devices ~ but then we come back to the same issue that [whisper](https://github.com/ethereum/whisper) and Status v1 had, i.e if we reduce the difficulty too much, then malicious actors with beefier machines can participate in spamming the network.

Reply:
The reasoning for 18 leading zeroes is mentioned in the Notion page. I can integrate it in here:

> Difficulty Level: The difficulty level is usually expressed as the number of leading zeros required in the hash output. It is chosen such that the computational effort required is significant but not prohibitive. We recommend a reasonable difficulty level that requires around 16-18 leading zeros in the SHA-256 hash. This would roughly take 0.65 to 2.62 seconds to compute in a low-grade CPU, capable of computing 100,000 hashes per second.

Reply:
Nice, thank you